timemachine/fe/interpolate.py

```diff
@@ -1,217 +1,138 @@
+from functools import partial
+from typing import Hashable
+
 import numpy as np
 
 
-class DuplicateIdxsError(RuntimeError):
+class DuplicateAlignmentKeysError(RuntimeError):
     pass
 
 
-def _add_dst_to_src_bond_or_angle(src_idxs, src_params, dst_idxs, dst_params):
+def align_idxs_and_params(
+    src_idxs,
+    src_params,
+    dst_idxs,
+    dst_params,
+    make_default,
+    key=lambda idxs, _: idxs,
+    get_idxs=lambda key: key,
+    validate_idxs=lambda _: None,
+):
     """
-    eg:
-    src_idxs: [[4,9], [3,4]]
-    src_params: [[a, b], [c, d]]
-
-    dst_idxs: [[3,4], [9,5]]
-    dst_params: [[e, f], [g, h]]
-
-    Generate end-state key value pairs
-
-    src_idxs: [[4,9], [3,4], [9,5]]
-    src_params: [[a,b], [c,d], [0,h]
-
+    Given two input parameter sets (idxs, params), aligns by the
+    specified key to produce two output parameter sets, where the
+    outputs have the same shape. When an (idxs, params) pair is
+    present in one input set but absent in the other, the missing
+    value is filled with a default computed from the value that is
+    present. By default, idxs are used as the alignment key.
+
+    Parameters
+    ----------
+    src_idxs, dst_idxs: array of int
+        Atom indices for each potential term. E.g. for harmonic bonds,
+        each would have shape (num_bonds, 2)
+    src_params, dst_params: array of float
+        Parameters corresponding to the specified indices
+    make_default: callable
+        Should return the value to fill for missing src (dst) given
+        the value present for dst (src)
+    key: callable
+        Should return the alignment key given arguments (idxs, params)
+    get_idxs: callable
+        Should return the idxs given an alignment key
+    validate_idxs: callable
+        Called on each set of idxs in src_idxs and dst_idxs; used to
+        validate input
+
+    Returns
+    -------
+    set
+        set of tuples (idxs, src_params, dst_params)
+
+    Examples
+    --------
+        >>> align_harmonic_angle_idxs_and_params = partial(align_idxs_and_params, make_default=lambda p: (0.0, p[1]))
+
+        For harmonic bonds, we align on idxs and fill missing
+        parameters with (0, d) where d is the equilibrium bond length
+
```

Comment:
and 0 is the corresponding force constant.

Reply:
Thanks, done in edcc4e3